compiler/src/dotty/tools/dotc/transform/SetRootTree.scala

```diff
@@ -3,41 +3,34 @@ package dotty.tools.dotc.transform
 import dotty.tools.dotc.CompilationUnit
 import dotty.tools.dotc.ast.tpd
 import dotty.tools.dotc.core.Contexts.Context
-import dotty.tools.dotc.transform.MegaPhase.MiniPhase
+import dotty.tools.dotc.core.Phases.Phase
 
 /** Set the `rootTreeOrProvider` property of class symbols. */
-class SetRootTree extends MiniPhase {
+class SetRootTree extends Phase {
 
   override val phaseName: String = SetRootTree.name
   override def isRunnable(implicit ctx: Context) =
     super.isRunnable && ctx.settings.YretainTrees.value
 
-  def runOn(unit: CompilationUnit)(implicit ctx: Context) = {
-    val runCtx = ctx.fresh.setCompilationUnit(unit)
-    traverser.traverse(unit.tpdTree)(runCtx)
-  }
-
-  override def transformTypeDef(tree: tpd.TypeDef)(implicit ctx: Context): tpd.Tree = {
-    if (tree.symbol.isClass) {
-      val sym = tree.symbol.asClass
-      tpd.sliceTopLevel(ctx.compilationUnit.tpdTree, sym) match {
-        case (pkg: tpd.PackageDef) :: Nil =>
-          sym.rootTreeOrProvider = pkg
-        case _ =>
-          sym.rootTreeOrProvider = tree
-      }
-    }
-    tree
+  override def run(implicit ctx: Context): Unit = {
+    val tree = ctx.compilationUnit.tpdTree
+    traverser.traverse(tree)
   }
 
   private def traverser = new tpd.TreeTraverser {
-    override def traverse(tree: tpd.Tree)(implicit ctx: Context): Unit = tree match {
-      case typeDef: tpd.TypeDef => transformTypeDef(typeDef)
-      case other => traverseChildren(other)
+    override def traverse(tree: tpd.Tree)(implicit ctx: Context): Unit = {
+      if (tree.isInstanceOf[tpd.TypeDef] && tree.symbol.isClass) {
+        val sym = tree.symbol.asClass
+        tpd.sliceTopLevel(ctx.compilationUnit.tpdTree, sym) match {
+          case (pkg: tpd.PackageDef) :: Nil =>
+            sym.rootTreeOrProvider = pkg
+          case _ =>
+            sym.rootTreeOrProvider = tree
+        }
+      }
+      traverseChildren(tree)
```

Comment:
Doesn't that mean that we would not be able to find the tree for an inner class directly?

With current master, for an inner class, we'd get the `TypeDef` if the source file needs to be `lateCompile`d, or the `PackageDef` otherwise.

With the change you suggest, we would always have an empty tree for inner classes. One would need to get the tree of the enclosing top level class and navigate from there.

Looking at `rootTreeOrProvider`, `rootTree` and `rootTreeContaining`, I'm under the impression that only `SourceTree` and `ReadTastyFromClasses` may need to be adapted. Can you think of something else that'd require changes?

Reply:
Actually, it looks like `rootTree` is only supposed to be called on top level class symbols, so everything should be fine, thanks.